File: excelium-core/src/main/java/excelium/core/reader/AbstractWorkbookReader.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.core.reader;

import excelium.common.ss.RangeLocation;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.StringUtils;

import java.io.Closeable;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Abstract workbook reader.
 * This class implements some helper methods for inquiring and searching cell values.
 *
 * @param <W> Workbook class
 * @param <S> Sheet class
 * @author PhungDucKien
 * @since 2018.04.23
 */
public abstract class AbstractWorkbookReader<W, S> implements WorkbookReader<W, S> {

    /**
     * Workbook object
     */
    protected W workbook;

    @Override
    public String findFirstOccurrence(Object value) throws IOException {
        List<S> sheets = listSheets();
        for (S sheet : sheets) {
            String sheetCellLocation = findFirstOccurrence(value, sheet);
            if (StringUtils.isNotBlank(sheetCellLocation)) {
                return sheetCellLocation;
            }
        }
        return null;
    }

    @Override
    public Map<Object, String> batchFindFirstOccurrence(List<Object> values) throws IOException {
        Map<Object, String> locationMap = new HashMap<>();

        List<S> sheets = listSheets();
        for (S sheet : sheets) {
            Map<Object, String> sheetCellLocations = batchFindFirstOccurrence(values, sheet);

            for (Object value : values) {
                if (!locationMap.containsKey(value) && StringUtils.isNotBlank(sheetCellLocations.get(value))) {
                    locationMap.put(value, sheetCellLocations.get(value));
                }
            }
        }
        return locationMap;
    }

    @Override
    public String findFirstOccurrence(Object value, S sheet) throws IOException {
        List<Object> values = new ArrayList<>();
        values.add(value);

        Map<Object, String> sheetCellLocations = batchFindFirstOccurrence(values, sheet);
        return sheetCellLocations.get(value);
    }

    @Override
    public Object getCellValue(String cell) throws IOException {
        return batchGetCellValues(cell).get(cell);
    }

    @Override
    public List<Object> getRowCellValues(String row) throws IOException {
        return batchGetRowCellValues(row).get(row);
    }

    @Override
    public List<Object> getColumnCellValues(String column) throws IOException {
        return batchGetColumnCellValues(column).get(column);
    }

    @Override
    public List<List<Object>> getRangeCellValues(String range) throws IOException {
        return batchGetRangeCellValues(range).get(range);
    }

    @Override
    public Map<String, Object> batchGetCellValues(String... cells) throws IOException {
        Map<String, List<List<Object>>> batchRangeCellValues = batchGetRangeCellValues(cells);
        Map<String, Object> values = new HashMap<>();
        for (String cell : batchRangeCellValues.keySet()) {
            List<List<Object>> rangeValue = batchRangeCellValues.get(cell);
            RangeLocation rangeLocation = new RangeLocation(cell);
            values.put(rangeLocation.formatAsString(), getFirstValue(rangeValue));
        }
        return values;
    }

    @Override
    public Map<String, List<Object>> batchGetRowCellValues(String... rows) throws IOException {
        Map<String, List<List<Object>>> batchRangeCellValues = batchGetRangeCellValues(rows);
        Map<String, List<Object>> values = new HashMap<>();
        for (String cell : batchRangeCellValues.keySet()) {
            List<List<Object>> rangeValue = batchRangeCellValues.get(cell);
            RangeLocation rangeLocation = new RangeLocation(cell);
            values.put(rangeLocation.formatAsString(), getFirstRowValues(rangeValue));
        }
        return values;
    }

    @Override
    public Map<String, List<Object>> batchGetColumnCellValues(String... columns) throws IOException {
        Map<String, List<List<Object>>> batchRangeCellValues = batchGetRangeCellValues(columns);
        Map<String, List<Object>> values = new HashMap<>();
        for (String cell : batchRangeCellValues.keySet()) {
            List<List<Object>> rangeValue = batchRangeCellValues.get(cell);
            RangeLocation rangeLocation = new RangeLocation(cell);
            values.put(rangeLocation.formatAsString(), getFirstColumnValues(rangeValue));
        }
        return values;
    }

    /**
     * Gets first value of 2-dimension list of objects
     *
     * @param values 2-dimension list of objects
     * @return first value
     */
    private Object getFirstValue(List<List<Object>> values) {
        if (CollectionUtils.isNotEmpty(values) && CollectionUtils.isNotEmpty(values.get(0))) {
            return values.get(0).get(0);
        }
        return null;
    }

    /**
     * Gets first row values of 2-dimension list of objects
     *
     * @param values 2-dimension list of objects
     * @return first row values
     */
    private List<Object> getFirstRowValues(List<List<Object>> values) {
        if (CollectionUtils.isNotEmpty(values)) {
            return values.get(0);
        }
        return new ArrayList<>();
    }

    /**
     * Gets first column values of 2-dimension list of objects
     *
     * @param values 2-dimension list of objects
     * @return first column values
     */
    private List<Object> getFirstColumnValues(List<List<Object>> values) {
        List<Object> columnValues = new ArrayList<>();
        if (CollectionUtils.isNotEmpty(values)) {
            for (List<Object> rowValues : values) {
                if (CollectionUtils.isNotEmpty(rowValues)) {
                    columnValues.add(rowValues.get(0));
                } else {
                    columnValues.add(null);
                }
            }
        }
        return columnValues;
    }

    @Override
    public void close() throws IOException {
        if (workbook instanceof Closeable) {
            ((Closeable) workbook).close();
        }
        workbook = null;
    }
}

File: packages/core/src/reader/AbstractWorkbookReader.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { CollectionUtil, RangeLocation, StringUtil } from '@excelium/common';
import WorkbookReader from './WorkbookReader';

/**
 * Abstract workbook reader.
 * This class implements some helper methods for inquiring and searching cell values.
 *
 * @param <W> Workbook class
 * @param <S> Sheet class
 * @author PhungDucKien
 * @since 2018.04.23
 */
export default abstract class AbstractWorkbookReader<W, S> implements WorkbookReader<W, S> {
  public abstract getWorkbook(): Promise<W>;

  public abstract getWorkbookName(workbook: W): string;

  public abstract listSheets(workbook: W): S[];

  public abstract getSheetName(sheet: S): string;

  public abstract batchFindFirstOccurrenceSheet(values: any[], sheet: S, workbook?: W): Promise<Map<string, string>>;

  public abstract batchGetRangeCellValues(ranges: string[], workbook: W): Promise<Map<string, any[][]>>;

  public async findFirstOccurrence(value: any, workbook: W): Promise<string | null> {
    const sheets = this.listSheets(workbook);
    for (const sheet of sheets) {
      const sheetCellLocation = await this.findFirstOccurrenceSheet(value, sheet, workbook);
      if (sheetCellLocation && StringUtil.isNotBlank(sheetCellLocation)) {
        return sheetCellLocation;
      }
    }
    return null;
  }

  public async batchFindFirstOccurrence(values: any[], workbook: W): Promise<Map<string, string>> {
    const locationMap = new Map<string, string>();

    const sheets = this.listSheets(workbook);
    for (const sheet of sheets) {
      const sheetCellLocations = await this.batchFindFirstOccurrenceSheet(values, sheet, workbook);

      for (const value of values) {
        const sheetCellLocation = sheetCellLocations.get(value);
        if (!locationMap.has(value) && sheetCellLocation && StringUtil.isNotBlank(sheetCellLocation)) {
          locationMap.set(value, sheetCellLocation);
        }
      }
    }
    return locationMap;
  }

  public async findFirstOccurrenceSheet(value: any, sheet: S, workbook?: W): Promise<string | null> {
    const values = [] as any[];
    values.push(value);

    const sheetCellLocations = await this.batchFindFirstOccurrenceSheet(values, sheet, workbook);
    return sheetCellLocations.get(value) || null;
  }

  public async getCellValue(cell: string, workbook: W): Promise<any> {
    return (await this.batchGetCellValues([cell], workbook)).get(cell);
  }

  public async getRowCellValues(row: string, workbook: W): Promise<any[]> {
    return (await this.batchGetRowCellValues([row], workbook)).get(row) || [];
  }

  public async getColumnCellValues(column: string, workbook: W): Promise<any[]> {
    return (await this.batchGetColumnCellValues([column], workbook)).get(column) || [];
  }

  public async getRangeCellValues(range: string, workbook: W): Promise<any[][]> {
    return (await this.batchGetRangeCellValues([range], workbook)).get(range) || [];
  }

  public async batchGetCellValues(cells: string[], workbook: W): Promise<Map<string, any>> {
    const batchRangeCellValues = await this.batchGetRangeCellValues(cells, workbook);
    const values = new Map<string, any>();
    batchRangeCellValues.forEach((rangeValue, cell) => {
      const rangeLocation = new RangeLocation({ reference: cell });
      values.set(rangeLocation.formatAsString(), this.getFirstValue(rangeValue));
    });
    return values;
  }

  public async batchGetRowCellValues(rows: string[], workbook: W): Promise<Map<string, any[]>> {
    const batchRangeCellValues = await this.batchGetRangeCellValues(rows, workbook);
    const values = new Map<string, any[]>();
    batchRangeCellValues.forEach((rangeValue, cell) => {
      const rangeLocation = new RangeLocation({ reference: cell });
      values.set(rangeLocation.formatAsString(), this.getFirstRowValues(rangeValue));
    });
    return values;
  }

  public async batchGetColumnCellValues(columns: string[], workbook: W): Promise<Map<string, any[]>> {
    const batchRangeCellValues = await this.batchGetRangeCellValues(columns, workbook);
    const values = new Map<string, any[]>();
    batchRangeCellValues.forEach((rangeValue, cell) => {
      const rangeLocation = new RangeLocation({ reference: cell });
      values.set(rangeLocation.formatAsString(), this.getFirstColumnValues(rangeValue));
    });
    return values;
  }

  /**
   * Gets first value of 2-dimension list of objects
   *
   * @param values 2-dimension list of objects
   * @return first value
   */
  public getFirstValue(values?: any[][]): any {
    if (values && CollectionUtil.isNotEmpty(values) && CollectionUtil.isNotEmpty(values[0])) {
      return values[0][0];
    }
    return null;
  }

  /**
   * Gets first row values of 2-dimension list of objects
   *
   * @param values 2-dimension list of objects
   * @return first row values
   */
  public getFirstRowValues(values?: any[][]): any[] {
    if (values && CollectionUtil.isNotEmpty(values)) {
      return values[0];
    }
    return [];
  }

  /**
   * Gets first column values of 2-dimension list of objects
   *
   * @param values 2-dimension list of objects
   * @return first column values
   */
  public getFirstColumnValues(values?: any[][]): any[] {
    const columnValues = [] as any[];
    if (values && CollectionUtil.isNotEmpty(values)) {
      for (const rowValues of values) {
        if (CollectionUtil.isNotEmpty(rowValues)) {
          columnValues.push(rowValues[0]);
        } else {
          columnValues.push(null);
        }
      }
    }
    return columnValues;
  }
}
